lib/modules.nix

```diff
@@ -52,6 +53,286 @@ in
+                  Whether to include this plugin in the `cmdline` completion source list.
+                '';
+              };
+              filetypes = lib.mkOption {
+                type = lib.types.attrsOf lib.types.bool;
+                # Only include `true` attrs in the final value
+                apply = lib.filterAttrs (_: lib.id);
+                default = enabledFiletypes;
+                # TODO: example
+                description = ''
+                  Whether to include this plugin in the specific `per_filetype` completion source lists.
+                '';
+              };
+              settings = lib.mkOption {
+                default = { };
+                description = ''
+                  Settings for the blink.cmp completion provider.
+                '';
+                example = settingsExample;
+                type = lib.types.submodule [
+                  {
+                    options = {
+                      enabled = internal;
+                      module = internal;
+                    };
+                    config = {
+                      name = lib.mkDefault sourceName;
+                      inherit module;
+                      # TODO: use mkDerivedConfig to preserve definition override priority?
+                      opts = lib.mkIf usePluginSettings pluginCfg.settings;
+                    };
+                  }
+                  (import ../plugins/by-name/blink-cmp/provider-config.nix lib)
+                ];
+              };
+            };
+          };
+        };
+      };
+      config = lib.mkIf (pluginCfg.enable && cfg.enable) {
+        plugins.blink-cmp.settings.sources = {
+          providers.${cfg.key} = cfg.settings;
+          default = lib.mkIf cfg.default [ cfg.key ];
+          # FIXME: the reference shows `cmdline` should/could be defined as a function
+          # https://cmp.saghen.dev/configuration/reference.html#sources
+          cmdline = lib.mkIf cfg.cmdline [ cfg.key ];
+          per_filetype = lib.mkIf (cfg.filetypes != { }) (
+            builtins.mapAttrs (_: _: [ cfg.key ]) cfg.filetypes
+          );
+        };
+      };
+    };
+
+  # Create a module configuring a plugin's integration with nvim-cmp and blink.cmp
+  mkCmpPluginModule =
+    {
+      # The plugin's option location-path
+      loc ? [
+        "plugins"
+        pluginName
+      ],
+      # Name of the plugin, used in documentation
+      pluginName,
+      # The nvim-cmp source name
+      # TODO: can we compute a sane default for sourceName?
+      sourceName,
+      # Defaults for the corresponding cmp options
+      enableDefault ? true,
+      enableCmdline ? false,
+      enabledFiletypes ? { },
+      # Whether to include a `blink` option at all
+      offerBlinkCompatibility ? true,
+      # Defaults for the blink compatibility option
+      enableBlinkProvider ? false,
+      enableBlinkDefault ? false,
+      enableBlinkCmdline ? false,
+      enabledBlinkFiletypes ? { },
```

Comment:
Should we just use the corresponding arguments above ?

Reply:
Not sure what you're asking. These are probably badly named, but they're intended to be the default values for the corresponding module options.

Reply:
Ah.

Because it felt unlikely that enabling a plugin as a nvim-cmp source by default should also enable it as a blink source by default.

We could have these defaults to the nvim-cmp equivalent, but I think it should still be possible to set these separately?